src/analyzer.fs

```diff
@@ -3,9 +3,155 @@
+
+    member this.using newContext go =
+        let old = this.varUse
+        this.varUse <- newContext
+        try
+            go()
+        finally
+            this.varUse <- old
+
+    // Vars in the AST can be a few things:
+    // [A] A declaration without initialization is neither a read nor a write.            int x;
+    // [B] A declaration with initialization is a "isDecl" write.                         int x = 1;
+    // [C] An assignment is a write.                                                      x = 1;
+    // [D] An augmented assignment is a read and then a write.                            x += 1;
+    // [E] A function name itself is neither a read nor a write.                          x();
+    // [F] A function call with an in parameter is a read.                                sin(x);
+    // [G] A function call with an out parameter is a write.                              void f(out p) { p = 1; } ... f(x);
+    // [H] A function call with an inout parameter is a read later followed by a write.   void f(inout p) { p += 1; } ... f(x);
+    //   Note that even when passing a global as out or inout, no aliasing happens.
+    //   The function works on a local copy, and copies back the value when exiting.
+    // [I] Other stray var uses are reads.
+
+    member this.visitExpr = function
+        | FunCall(Op op as fct, first::args) when Builtin.assignOps.Contains op ->
+            let alsoReadTheVar = op <> "=" // Augmented assignment or ++ or --
+            if alsoReadTheVar then
+                // Visit the lhs of the assignment as a read. [D]
+                this.using { this.varUse with access.isWrite = false; access.isRead = true } (fun () ->
+                    this.visitExpr first
+                )
+            List.iter this.visitExpr args // visit the rhs of the assignments
+            this.visitExpr fct // visit the assignment op
+            // Visit the lhs of the assignment as a write. [C] [D]
+            this.using { this.varUse with access.isWrite = true; access.isRead = false } (fun () ->
+                this.visitExpr first
+            )
+        | FunCall(fct, args) ->
+            this.using { this.varUse with access.isWrite = false; access.isRead = false } (fun () -> // [E]
+                this.visitExpr fct
+            )
+            // Handle in/out/inout parameters.
+            let funcDecl =
+                match fct with 
+                | Var v -> 
+                    match v.Declaration with
+                    | Declaration.UserFunction funcDecl -> Some funcDecl
+                    | _ -> None
+                | _ -> None
+            let paramAccessList =
+                funcDecl
+                |> Option.map (fun funcDecl -> funcDecl.funcType.args |> List.map (fun (ty, _) -> ty.access))
+                |> Option.defaultWith (fun () -> args |> List.map (fun _ -> { isWrite = false; isRead = true }))
```

Comment:
opt. might be clearer to do a `match funcDecl` instead of Option.map + Option.defaultWith